File: p001e/two_sum.cpp
// Build program by
// g++ -std=c++11 *.cpp -o test && ./test && rm test
#include <cassert>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target)
    {
        unordered_map<int, int> seen;
        for (int i = 0; i < nums.size(); ++i) {
            int key = target - nums[i];
            if (seen.count(key) != 0) {
                return { seen[key], i };
            } else {
                seen[nums[i]] = i;
            }
        }
        return vector<int>();
    }
};

// TEST
struct Test {
    vector<int> nums;
    int target;
    vector<int> expected;
    void run()
    {
        Solution sol;
        auto actual = sol.twoSum(nums, target);
        cout << "Two sum in [";
        for (auto n : nums) {
            cout << n << ", ";
        }
        cout << "] for target " << target << " -> ";
        cout << "[" << actual[0] << ", " << actual[1] << "]" << endl;
        assert(actual[0] == expected[0] && actual[1] == expected[1]);
    }
};

int main(int argc, char* argv[])
{
    vector<Test> tests = {
        { { 2, 7, 11, 15 }, 9, { 0, 1 } },
        { { 3, 2, 4 }, 6, { 1, 2 } },
        { { 3, 3 }, 6, { 0, 1 } },
    };
    for (auto t : tests)
        t.run();
    return 0;
}

File: p001e/two_sum.py
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen = dict()
        for i, n in enumerate(nums):
            if target - n in seen:
                return [seen[target - n], i]
            else:
                seen[n] = i
        return [-1, -1]


# TESTS
for nums, target, expected in [
    ([2, 7, 11, 15], 9, [0, 1]),
    ([3, 2, 4], 6, [1, 2]),
    ([3, 3], 6, [0, 1]),
]:
    sol = Solution()
    actual = sol.twoSum(nums, target)
    print("Two sum in", nums, "for target", target, "->", actual)
    assert actual == expected
